tests/custom_function_test.py

```diff
@@ -22,3 +22,22 @@ def test_ternary(self):
         expression = jsonata.Jsonata("$abc(a,b,c)")
         expression.register_lambda("abc", lambda x, y, z: str(x) + str(y) + str(z))
         assert expression.evaluate({"a": "a", "b": "b", "c": "c"}) == "abc"
+
+    def test_map_with_lambda(self):
+        expression = jsonata.Jsonata("$map([1, 2, 3], $square)")
+        expression.register_lambda("square", lambda x: x * x)
+        assert expression.evaluate(None) == [1, 4, 9]
+
+    def test_map_with_function(self):
+        expression = jsonata.Jsonata("$map([1, 2, 3], function($v) { $v * $v })")
+        assert expression.evaluate(None) == [1, 4, 9]
+
+    def test_singleton_map(self):
+        expression = jsonata.Jsonata("$map([1], $square)[]")
+        expression.register_lambda("square", lambda x: x * x)
+        assert expression.evaluate(None) == [1]
+
+    def test_singleton_map_with_chain(self):
+        expression = jsonata.Jsonata("$ ~> $map($square)[]")
```

Comment:
When I try the following in jsonata.js, it fails.  Perhaps this is a bug in jsonata.js as well (unless I'm doing something wrong)?  If so, I'll wait until it is fixed there before making the fix here.

```
            var expr = jsonata("$ ~> $map($square)[]");
            expr.registerFunction("square", function(x) {
                return x * x;
            });
            var result = await expr.evaluate([1]);
            var expected = [1];
            expect(result).to.deep.equal(expected);
```

Reply:
indeed, I did not test the behavior of the original jsonata.js implementation. Your observation is correct, that also does not work there and the problem seems to be that the chain operator does not correctly propagate whether any child expression is forced to keep the array.

Will report the problem there with a reference to this ticket. Probably the same fix needs to be applied there.

However I want to express that without this working, the map function is pretty much useless to me in my use-case. The workaround is to always create a lambda function myself to operate on all values of an array, which is a bit of a pain. I dont see any value in a map function that returns different results depending on the input values.